src/Params/WaveTable.h

```diff
@@ -0,0 +1,659 @@
+            m_data = new T[base_type::size()];
+    }
+public:
+    Tensor(tensor_size_t size) : TensorBase<1, T>(size),
+        m_data(size ? new T[size] : nullptr) {}
+    Tensor(const Shape<1>& shape) : Tensor(shape.dim[0]){}
+    ~Tensor() { delete[] m_data; }
+    Tensor& operator=(Tensor&& other) {
+        base_type::operator=(other);
+        m_data = other.m_data;
+    }
+
+    //! raw access into 1D array
+    T* data() { return m_data; }
+    const T* data() const { return m_data; }
+    //! raw indexing into 1D array
+    T& operator[](tensor_size_t i) { return m_data[i]; }
+    const T& operator[](tensor_size_t i) const { return m_data[i]; }
+
+    bool operator==(const Tensor<1, T>& other) const {
+        return base_type::operator==(other) &&
+            std::equal(m_data, m_data + base_type::size(), other.m_data);
+    }
+
+    bool operator!=(const Tensor<1, T>& other) const {
+        return !operator==(other); }
+
+    void swapWith(Tensor<1,T>& other)
+    {
+        TensorBase<1, T>::swapWith(other);
+        std::swap(m_data, other.m_data);
+    }
+
+    void swapDataWith(T*& new_data)
+    {
+        std::swap(m_data, new_data);
+    }
+
+    class TensorIterator
+    {
+        T* sub;
+    public:
+        // define all the stuff that std::lower_bound & co require...
+        using difference_type = std::size_t;
+        using value_type = T;
+        using pointer = T*;
+        using reference = T&;
+        using iterator_category = std::random_access_iterator_tag;
+
+        TensorIterator() = default;
+        TensorIterator(T* sub) : sub(sub) {}
+        TensorIterator& operator+=(tensor_size_t inc) { sub += inc; return *this; }
+        TensorIterator operator-(tensor_size_t inc) { return TensorIterator(sub - inc); }
+        TensorIterator& operator++() { ++sub; return *this; }
+        TensorIterator& operator--() { --sub; return *this; }
+        bool operator==(const TensorIterator& other) const { return sub == other.sub; }
+        bool operator!=(const TensorIterator& other) const { return sub != other.sub; }
+        tensor_size_t operator-(const TensorIterator& other) const { return sub - other.sub; }
+        T& operator*() { return *sub; }
+    };
+    TensorIterator begin() const { return TensorIterator(m_data); }
+    TensorIterator end() const { return TensorIterator(m_data + base_type::size()); }
+
+    /*
+        testing only
+     */
+    tensor_size_t debug_set_data_using_deep_copy(const T* new_data)
+    {
+        std::copy(new_data, new_data+base_type::size(), m_data);
+        return base_type::size();
+    }
+
+    Shape<1> debug_get_shape() const { return Shape<1>{base_type::size()}; }
+};
+
+/*
+ * Ringbuffer
+ */
+
+/**
+    Ringbuffer without buffer. Only size, reader and writer
+    (and delayed writer).
+    @invariant r <= w_delayed <= w (non-cyclic)
+    @invariant r <= w + m_size (non-cyclic)
+
+    While "normal" ringbuffers only allow w < r + m_size, this ringbuffer
```

Comment:
aha, this looks like this is going to be the explanation I was asking about earlier